linalg/densemat.cpp

```diff
@@ -3503,4 +3505,145 @@ DenseTensor &DenseTensor::operator=(double c)
    return *this;
 }
 
+void BatchLUFactor(Vector &Minv, const int m, const int NE, Array<int> &P)
+{
+   P.SetSize(m*NE);
+   BatchLUFactor_impl(Minv.ReadWrite(), m, NE, P.Write());
+}
+
+void BatchLUFactor(DenseTensor &Minv, Array<int> &P)
+{
+   const int m = Minv.SizeI();
+   const int NE = Minv.SizeK();
+   P.SetSize(m*NE);
+   BatchLUFactor_impl(Minv.ReadWrite(), m, NE, P.Write());
+}
+
+void BatchLUFactor_impl(double *Minv, const int m, const int NE, int *P)
+{
+
+   auto data_all = mfem::Reshape(Minv, m, m, NE);
+   auto piv_all = mfem::Reshape(P, m, NE);
+   Array<bool> pivot_flag(1);
+   pivot_flag[0] = true;
+   bool *d_pivot_flag = pivot_flag.ReadWrite();
+   const double TOL = 1.e-9;
+
+   MFEM_FORALL(e, NE,
+   {
+
+      double *data = &data_all(0,0,e);
+      int *ipiv = &piv_all(0,e);
+      for (int i = 0; i < m; i++)
+      {
+
+         // pivoting
+         {
+            int piv = i;
+            double a = fabs(data[piv+i*m]);
+            for (int j = i+1; j < m; j++)
+            {
+               const double b = fabs(data[j+i*m]);
+               if (b > a)
+               {
+                  a = b;
+                  piv = j;
+               }
+            }
+            ipiv[i] = piv;
+            if (piv != i)
+            {
+               // swap rows i and piv in both L and U parts
+               for (int j = 0; j < m; j++)
+               {
+                  mfem::kernels::internal::Swap<double>(data[i+j*m], data[piv+j*m]);
+               }
+            }
+         }//pivot end
+
+         if (abs(data[i + i*m]) <= TOL)
+         {
+            d_pivot_flag[0] = false;
+         }
+
+         const double a_ii_inv = 1.0 / data[i+i*m];
+         for (int j = i+1; j < m; j++)
+         {
+            data[j+i*m] *= a_ii_inv;
+         }
+
+         for (int k = i+1; k < m; k++)
+         {
+            const double a_ik = data[i+k*m];
+            for (int j = i+1; j < m; j++)
+            {
+               data[j+k*m] -= a_ik * data[j+i*m];
+            }
+         }
+
+      }//m loop
+
+   });
+
+   MFEM_ASSERT(pivot_flag[0].HostRead(), "Batch LU factorization failed \n");
```

Comment:
@v-dobrev , I think this should do the trick. Let me know if this works for you?